File: personal_library/lib/services/user_service.dart
import 'dart:io';
import 'package:dio/dio.dart';
import '../core/network/api_client.dart';
import '../models/user_model.dart';

class UserService {
  // ─── LẤY THÔNG TIN CÁ NHÂN ──────────────────────
  Future<UserModel> getProfile() async {
    try {
      final response = await ApiClient.dio.get('/users/profile');
      return UserModel.fromJson(response.data['data']);
    } catch (e) {
      throw Exception('Không thể lấy thông tin: $e');
    }
  }

  // ─── CẬP NHẬT THÔNG TIN ─────────────────────────
  Future<UserModel> updateProfile({
    required String tenHienThi,
    String? anhDaiDien,
  }) async {
    try {
      final data = <String, dynamic>{
        'ten_hien_thi': tenHienThi,
      };
      if (anhDaiDien != null && anhDaiDien.isNotEmpty) {
        data['anh_dai_dien'] = anhDaiDien;
      }

      final response = await ApiClient.dio.put('/users/profile', data: data);
      return UserModel.fromJson(response.data['data']);
    } catch (e) {
      throw Exception('Không thể cập nhật: $e');
    }
  }

  // ─── ĐỔI MẬT KHẨU ─────────────────────────────────
  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
  }) async {
    try {
      await ApiClient.dio.post('/users/change-password', data: {
        'currentPassword': currentPassword,
        'newPassword': newPassword,
      });
    } catch (e) {
      throw Exception('Không thể đổi mật khẩu: $e');
    }
  }

  // ─── UPLOAD ẢNH ĐẠI DIỆN ─────────────────────────
  Future<String> uploadAvatar(File imageFile) async {
    try {
      final formData = FormData.fromMap({
        'avatar': await MultipartFile.fromFile(imageFile.path),
      });
      final response = await ApiClient.dio.post(
        '/users/upload-avatar',
        data: formData,
        options: Options(
          headers: {'Content-Type': 'multipart/form-data'},
        ),
      );
      return response.data['data']['avatar_url'];
    } catch (e) {
      throw Exception('Không thể upload ảnh: $e');
    }
  }


Future<List<UserModel>> getAllUsers() async {
  try {
    print('📤 ===== UserService.getAllUsers =====');
    
    final response = await ApiClient.dio.get('/users');
    print('📥 Status: ${response.statusCode}');
    print('📥 Data: ${response.data}');
    
    final data = response.data['data'] as List? ?? [];
    print('📥 Data length: ${data.length}');
    
    final users = data.map((e) => UserModel.fromJson(e)).toList();
    print('📥 Users parsed: ${users.length}');
    print('📥 First user: ${users.isNotEmpty ? users[0].tenHienThi : 'empty'}');
    
    return users;
  } catch (e) {
    print('❌ getAllUsers error: $e');
    if (e is DioException) {
      print('❌ Response: ${e.response?.data}');
      print('❌ Status: ${e.response?.statusCode}');
    }
    throw Exception('Không thể lấy danh sách: $e');
  }
}

  // ─── CẬP NHẬT VAI TRÒ (CHỈ THU_THU) ──────────────
  Future<UserModel> updateUserRole({
    required String userId,
    required String vaiTro,
  }) async {
    try {
      final response = await ApiClient.dio.patch(
        '/users/$userId/role',
        data: {'vai_tro': vaiTro},
      );
      return UserModel.fromJson(response.data['data']);
    } catch (e) {
      throw Exception('Không thể cập nhật vai trò: $e');
    }
  }

  // ─── XÓA NGƯỜI DÙNG (CHỈ THU_THU) ────────────────
  Future<void> deleteUser(String userId) async {
    try {
      await ApiClient.dio.delete('/users/$userId');
    } catch (e) {
      throw Exception('Không thể xóa người dùng: $e');
    }
  }
}

File: backend/src/services/user.service.js
// services/user.service.js
const prisma = require('../config/prisma');
const bcrypt = require('bcrypt');

// ================= GET USER BY ID =================
exports.getUserById = async (userId) => {
    return await prisma.nguoi_dung.findUnique({
        where: { id: userId },
        select: {
            id: true,
            email: true,
            ten_hien_thi: true,
            anh_dai_dien: true,
            vai_tro: true,
            trang_thai: true,
            created_at: true,
            updated_at: true
        }
    });
};

// ================= UPDATE PROFILE =================
exports.updateProfile = async (userId, data) => {
    return await prisma.nguoi_dung.update({
        where: { id: userId },
        data: {
            ten_hien_thi: data.ten_hien_thi,
            anh_dai_dien: data.anh_dai_dien,
            updated_at: new Date()
        },
        select: {
            id: true,
            email: true,
            ten_hien_thi: true,
            anh_dai_dien: true,
            vai_tro: true
        }
    });
};

// ================= CHANGE PASSWORD =================
exports.changePassword = async (userId, currentPassword, newPassword) => {
    const user = await prisma.nguoi_dung.findUnique({
        where: { id: userId },
        select: { mat_khau_hash: true }
    });

    if (!user) {
        throw new Error('Không tìm thấy người dùng');
    }

    const isMatch = await bcrypt.compare(currentPassword, user.mat_khau_hash);
    if (!isMatch) {
        throw new Error('Mật khẩu hiện tại không đúng');
    }

    const hashedPassword = await bcrypt.hash(newPassword, 10);

    await prisma.nguoi_dung.update({
        where: { id: userId },
        data: {
            mat_khau_hash: hashedPassword,
            updated_at: new Date()
        }
    });
};

// ================= GET ALL USERS =================
exports.getAllUsers = async () => {  // 🔥 KHÔNG CÓ req, res
    console.log('🔍 getAllUsers called');
    
    const users = await prisma.nguoi_dung.findMany({
        select: {
            id: true,
            email: true,
            ten_hien_thi: true,
            anh_dai_dien: true,
            vai_tro: true,
            trang_thai: true,
            created_at: true,
            updated_at: true
        },
        orderBy: { created_at: 'desc' }
    });
    
    console.log(`✅ Found ${users.length} users`);
    return users;
};

// ================= UPDATE USER ROLE =================
exports.updateUserRole = async (userId, vaiTro) => {
    return await prisma.nguoi_dung.update({
        where: { id: userId },
        data: {
            vai_tro: vaiTro,
            updated_at: new Date()
        },
        select: {
            id: true,
            ten_hien_thi: true,
            vai_tro: true
        }
    });
};

// ================= DELETE USER =================
exports.deleteUser = async (userId) => {
    return await prisma.nguoi_dung.delete({
        where: { id: userId }
    });
};